base/sysinfo.jl

```diff
@@ -543,9 +543,20 @@ function which(program_name::String)
     for path_dir in path_dirs
         for pname in program_names
             program_path = joinpath(path_dir, pname)
-            # If we find something that matches our name and we can execute
-            if isfile(program_path) && isexecutable(program_path)
-                return program_path
+            try
+                # If we find something that matches our name and we can execute
+                if isfile(program_path) && isexecutable(program_path)
+                    return program_path
+                end
+            catch e
+                # If we encounter a permission error, we skip this directory
+                # and continue to the next directory in the PATH variable.
+                if isa(e, IOError) && e.code == Base.UV_EACCES
```

Comment:
Is there some way to create a test to exercise this code path? Maybe we create two temporary directories and put “foo” inside them both, give one directory zero permissions, add it to the front of PATH, then ensure that Sys.which(“foo”) works as expected.

Reply:
This should use the `access` syscall, not the `isexecutable` metadata (which, despite the name, doesn't tell you if you can execute it)

Reply:
I think the issue is this:
```
julia> Sys.isexecutable(mktempdir())
true
```

Reply:
#49181 has a reproducer which is a mix of shell commands and julia code, but it shouldn't be hard to write everything in Julia.